File: scripts/core/coverage.py
import json


def calculate_factual_confidence(conn, baseline_files, baseline_actions, output_topics):
    if not baseline_files and not baseline_actions:
        return 1.0
    covered_files = 0
    covered_actions = 0
    decisions_text = ""
    for t in output_topics:
        for d in t.get("decisions", []):
            decisions_text += " " + d.get("decision", "").lower() + " " + d.get("rationale", "").lower()

    for f in baseline_files:
        if f.lower() in decisions_text:
            covered_files += 1

    for action in baseline_actions:
        if action.startswith("confirm:"):
            dec_id = action.split(":")[1]
            try:
                cursor = conn.execute("SELECT user_confirmed FROM topic_decisions WHERE id=?", (dec_id,))
                row = cursor.fetchone()
                if row and row[0] == 1:
                    covered_actions += 1
            except Exception:
                pass

    total_items = len(baseline_files) + len(baseline_actions)
    covered_items = covered_files + covered_actions
    return min(1.0, covered_items / total_items) if total_items > 0 else 1.0


def validate_id_inheritance(conn, project_uuid, new_topics):
    cursor = conn.execute(
        "SELECT id FROM topic_decisions WHERE project_uuid = ? AND user_confirmed = 1",
        (project_uuid,)
    )
    confirmed_ids = {row[0] for row in cursor.fetchall()}
    if not confirmed_ids:
        return True
    inherited_ids = set()
    for t in new_topics:
        for d in t.get("decisions", []):
            for val in d.get("inherited_from", []):
                try:
                    inherited_ids.add(int(val))
                except (ValueError, TypeError):
                    pass
    missing_ids = confirmed_ids - inherited_ids
    if missing_ids:
        print(f"REMORA HARD ANCHOR VIOLATION WARNING: user_confirmed=1 IDs lost: {list(missing_ids)}.")
    return True

File: packages/core/src/coverage.ts
import Database from "better-sqlite3";
import { getDecisionConfirmed, getConfirmedDecisionIds } from "./storage/decisions";

interface Decision {
  decision?: string;
  rationale?: string;
  inherited_from?: unknown[];
}

interface Topic {
  decisions?: Decision[];
}

/**
 * Calculate a confidence score (0-1) based on how many baseline files and actions
 * are covered by the output topics' decision text and confirmed decision IDs.
 */
export function calculateFactualConfidence(
  baselineFiles: string[],
  baselineActions: string[],
  outputTopics: Topic[],
  conn?: Database
): number {
  if (!baselineFiles.length && !baselineActions.length) {
    return 1.0;
  }

  let coveredFiles = 0;
  let coveredActions = 0;
  let decisionsText = "";

  for (const t of outputTopics) {
    for (const d of t.decisions ?? []) {
      decisionsText +=
        " " +
        (d.decision ?? "").toLowerCase() +
        " " +
        (d.rationale ?? "").toLowerCase();
    }
  }

  for (const f of baselineFiles) {
    if (decisionsText.includes(f.toLowerCase())) {
      coveredFiles += 1;
    }
  }

  for (const action of baselineActions) {
    if (action.startsWith("confirm:")) {
      const decId = action.split(":")[1];
      const decIdNum = parseInt(decId, 10);
      if (!isNaN(decIdNum)) {
        try {
          coveredActions += getDecisionConfirmed(decIdNum, conn) ? 1 : 0;
        } catch {
          // pass
        }
      }
    }
  }

  const totalItems = baselineFiles.length + baselineActions.length;
  const coveredItems = coveredFiles + coveredActions;
  return totalItems > 0 ? Math.min(1.0, coveredItems / totalItems) : 1.0;
}

/**
 * Validate that all confirmed decision IDs from the database are still present
 * in the new topics' `inherited_from` fields.
 *
 * Logs a warning if any confirmed IDs are missing (hard anchor violation).
 */
export function validateIdInheritance(
  projectUuid: string,
  newTopics: Topic[],
  conn?: Database
): boolean {
  const confirmedIds = getConfirmedDecisionIds(projectUuid, conn);
  if (!confirmedIds || confirmedIds.size === 0) {
    return true;
  }

  const inheritedIds = new Set<number>();
  for (const t of newTopics) {
    for (const d of t.decisions ?? []) {
      for (const val of d.inherited_from ?? []) {
        const num = parseInt(String(val), 10);
        if (!isNaN(num)) {
          inheritedIds.add(num);
        }
      }
    }
  }

  const missingIds = new Set(
    [...confirmedIds].filter((x) => !inheritedIds.has(x))
  );
  if (missingIds.size > 0) {
    console.log(
      `REMORA HARD ANCHOR VIOLATION WARNING: user_confirmed=1 IDs lost: ${[
        ...missingIds,
      ]}.`
    );
  }
  return true;
}
